{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import agentpy as ap\n",
    "import random\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from IPython.display import Image\n",
    "import networkx as nx\n",
    "from agentpy.tools import make_list\n",
    "\n",
    "# For using ggplot from r for better figures \n",
    "from pyper import *\n",
    "r = R(RCMD=\"/Library/Frameworks/R.framework/Resources/bin/R\", use_pandas=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Python implementation of the agent-based model from \"Sex specific kinship dynamics and the fitness consequences of the social environment in killer whales\". The model captures social network dynamics and the evolution of helping and harming behaviour."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [],
   "source": [
    "class AnimalAgent(ap.Agent):\n",
    "\n",
    "    def setup(self):\n",
    "        \"\"\" Initialise new agent variables at agent creation \"\"\"\n",
    "        self.age = np.random.geometric(1/self.p.N) # *** Why does the rate change with N?\n",
    "        self.ageClass = self.set_age_class()\n",
    "        self.sex = random.choice(['m','f'])\n",
    "        # Random starting values. np array m/f rows (index 0/1), immature/adult (0/1) cols\n",
    "        self.genome = np.reshape(np.random.choice(['c','d'],4), (2,2))\n",
    "        self.fitness = 0\n",
    "\n",
    "    def set_age_class(self):\n",
    "        \"\"\" Set the age class based on a threshold \"\"\"\n",
    "        if self.age < self.p.m:\n",
    "            self.ageClass = 'immature'\n",
    "        else:\n",
    "            self.ageClass = 'adult'\n",
    "\n",
    "    def increase_age(self):\n",
    "        self.age += 1\n",
    "        self.set_age_class()\n",
    "\n",
    "    def get_genome_from_parents(self, parent1_genome, parent2_genome):\n",
    "        \"\"\" Uniform crossover between the two parents to return the child's genome \"\"\"\n",
    "        choice = np.random.randint(2, size = parent1_genome.size).reshape(parent1_genome.shape).astype(bool)\n",
    "        self.genome = np.where(choice, parent1_genome, parent2_genome)\n",
    "\n",
    "    def mutate_genome(self):\n",
    "        \"\"\" Random uniform mutation of each gene \"\"\"\n",
    "\n",
    "        # Create numpy mask to select each gene to mutate with given probability\n",
    "        probs = [self.p.mut_prob, 1-self.p.mut_prob]\n",
    "        mask = np.random.choice([True, False], self.genome.shape, p=probs)\n",
    "        # slip the values for those chosen to mutate\n",
    "        self.genome[mask] = np.where(self.genome=='d', 'c', 'd')[mask]\n",
    "\n",
    "    def social_interaction(self):\n",
    "        \"\"\" The agent plays the cooperation game with each social partner \"\"\"\n",
    "        # Get the angent's strategy\n",
    "        # m/f rows (index 0/1), immature/adult cols (0/1)\n",
    "        row = 0 if self.sex == 'm' else 1\n",
    "        col = 0 if self.ageClass == 'immature' else 1 \n",
    "        strategy = self.genome[row,col]\n",
    "        if strategy == 'c':\n",
    "            self.fitness -= self.p.cost\n",
    "            for n in self.network.neighbors(self):\n",
    "                n.fitness += self.p.benefit\n",
    "        else:\n",
    "            return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [],
   "source": [
    "class MainModel(ap.Model):\n",
    "\n",
    "    def setup(self):\n",
    "        \"\"\" Initialise the model and setup the agent population \"\"\"\n",
    "\n",
    "        # Create agent population\n",
    "        self.pop = ap.AgentDList(self, self.p.N, AnimalAgent)\n",
    "\n",
    "        # Create a random undirected network with desired mean degree\n",
    "        max_edges = (self.p.N * (self.p.N - 1) / 2)\n",
    "        desired_edges = (self.p.N * self.p.d) / 2\n",
    "        prob_edge = desired_edges / max_edges\n",
    "        graph = nx.erdos_renyi_graph(n=self.p.N, p=prob_edge)\n",
    "        # Define the network in the model and give agents reference to it\n",
    "        self.network = self.pop.network = ap.Network(self, graph)\n",
    "        self.network.add_agents(self.pop, self.network.nodes)\n",
    "    \n",
    "    def kill_agent(self, agent_to_kill):\n",
    "        \"\"\" Removes the agent and node from the network and deletes agent \"\"\"\n",
    "\n",
    "        # Get the node the agent is on and remove the node\n",
    "        agent_to_kill_node = self.network.positions[agent_to_kill]\n",
    "        self.network.remove_node(agent_to_kill_node)\n",
    "        # Remove agent from population\n",
    "        self.pop.remove(agent_to_kill)\n",
    "\n",
    "    def birth_agent(self, sex, mother_agent, father_agent):\n",
    "        \"\"\" Creates a single new agent and sets it up in the network\n",
    "            New agents inherit each gene at random from either parent \"\"\"\n",
    "\n",
    "        # Add new agent to the population\n",
    "        new_agent = ap.AgentDList(self, 1, AnimalAgent)\n",
    "        new_agent[0].sex = sex\n",
    "        new_agent[0].get_genome_from_parents(mother_agent.genome, father_agent.genome)\n",
    "        new_agent[0].mutate_genome()\n",
    "        self.pop += new_agent\n",
    "        # Set up the agent on a new node on the network\n",
    "        self.network.add_agents(new_agent)\n",
    "        # Give the agent a connection to its mother\n",
    "        agent_node = self.network.positions[new_agent[0]]\n",
    "        mother_node = self.network.positions[mother_agent]\n",
    "        self.network.graph.add_edge(agent_node, mother_node)\n",
    "        # Give it D random new associates (keeping the mean degree d on ave)\n",
    "        D = np.random.binomial(self.p.N-2, (self.p.d-1)/(self.p.N-2))\n",
    "        # Individuals connected to the mother are w times more likely to be selected\n",
    "        mother_neighbors = list(model.network.neighbors(model.pop[0]))\n",
    "        mother_neighbor_nodes = [model.network.positions[node] for node in mother_neighbors]\n",
    "        connected_to_mother = [model.network.positions[agt] in mother_neighbor_nodes for agt in model.pop]\n",
    "        probs = [self.p.w if connected else 1 for connected in connected_to_mother]\n",
    "        probs = [ p/sum(probs) for p in probs]\n",
    "        for i in range(D):\n",
    "            rand_node = np.random.choice(list(self.network.graph.nodes),p=probs)\n",
    "            # Check that associate isn't self or already connected\n",
    "            while (rand_node==agent_node) or (self.network.graph.has_edge(agent_node, rand_node)):\n",
    "                rand_node = np.random.choice(list(self.network.graph.nodes),p=probs)\n",
    "            self.network.graph.add_edge(agent_node, rand_node)\n",
    "\n",
    "    def select_agent_to_kill(self):\n",
    "        \"\"\" Returns which agent to kill.\n",
    "        Works by selecting agents to survive and seeing who is left\n",
    "        \"\"\"\n",
    "        # Make a list of agent fitnesses aligned with the agents in order\n",
    "        # Make another list \n",
    "        \n",
    "        return( self.pop.select(self.pop.agent_alive == False) )\n",
    "\n",
    "    def update(self):\n",
    "        \"\"\" Called after setup and every step \"\"\"\n",
    "\n",
    "        # Record stats\n",
    "        # self.pop.record(\"opinion\")\n",
    "\n",
    "    def step(self):\n",
    "        \"\"\" Model events per timestep \"\"\"\n",
    "\n",
    "        # self.pop.shuffle().opinion_influence()\n",
    "        self.pop.social_interaction()\n",
    "        #agent_to_kill = self.select_agent_to_kill()\n",
    "        #self.kill_agent(agent_to_kill)\n",
    "        #birth agent **********************\n",
    "        self.pop.increase_age()\n",
    "\n",
    "    def end(self):\n",
    "        \"\"\" Termination conditions \"\"\"\n",
    "        return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [],
   "source": [
    "parameters = {\n",
    "    'N': 20,         # population size ********* What is this?\n",
    "    'steps': 100,    # Generations\n",
    "    'm': 10,         # Age of maturity ********************* WE DON'T KNOW WHAT THIS IS\n",
    "    'w': 28,         # times more likely offspring select mothers' social partner over others\n",
    "    'd': 19.42,      # Mean degree of the social network\n",
    "    'mut_prob': 0.1, # Mutation probability\n",
    "    'benefit': 1,\n",
    "    'cost': 0.1,\n",
    "    'fitness_baseline': 0.2\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Completed: 100 steps\n",
      "Run time: 0:00:00.059698\n",
      "Simulation finished\n"
     ]
    }
   ],
   "source": [
    "# Single run of the model (experiments later)\n",
    "model = MainModel(parameters)\n",
    "results = model.run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "variables",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m/opt/anaconda3/envs/ABM_env/lib/python3.10/site-packages/agentpy/tools.py:87\u001b[0m, in \u001b[0;36mAttrDict.__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m     <a href='file:///opt/anaconda3/envs/ABM_env/lib/python3.10/site-packages/agentpy/tools.py?line=85'>86</a>\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m---> <a href='file:///opt/anaconda3/envs/ABM_env/lib/python3.10/site-packages/agentpy/tools.py?line=86'>87</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m\u001b[39m__getitem__\u001b[39;49m(name)\n\u001b[1;32m     <a href='file:///opt/anaconda3/envs/ABM_env/lib/python3.10/site-packages/agentpy/tools.py?line=87'>88</a>\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mKeyError\u001b[39;00m:\n\u001b[1;32m     <a href='file:///opt/anaconda3/envs/ABM_env/lib/python3.10/site-packages/agentpy/tools.py?line=88'>89</a>\u001b[0m     \u001b[39m# Important for pickle to work\u001b[39;00m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'variables'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[1;32m/Users/df525/code/weiss_helping_harming2022/helping_harming_model.ipynb Cell 17'\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/df525/code/weiss_helping_harming2022/helping_harming_model.ipynb#ch0000010?line=0'>1</a>\u001b[0m \u001b[39m# Get all logged data for agents\u001b[39;00m\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/df525/code/weiss_helping_harming2022/helping_harming_model.ipynb#ch0000010?line=1'>2</a>\u001b[0m results \u001b[39m=\u001b[39m results\u001b[39m.\u001b[39;49mvariables\u001b[39m.\u001b[39mAnimalAgent\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/df525/code/weiss_helping_harming2022/helping_harming_model.ipynb#ch0000010?line=2'>3</a>\u001b[0m \u001b[39m# Convert the resulting Pandas series to a dataFrame\u001b[39;00m\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/df525/code/weiss_helping_harming2022/helping_harming_model.ipynb#ch0000010?line=3'>4</a>\u001b[0m results \u001b[39m=\u001b[39m results\u001b[39m.\u001b[39mgenome\u001b[39m.\u001b[39mto_frame()\u001b[39m.\u001b[39mreset_index()\n",
      "File \u001b[0;32m/opt/anaconda3/envs/ABM_env/lib/python3.10/site-packages/agentpy/tools.py:90\u001b[0m, in \u001b[0;36mAttrDict.__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m     <a href='file:///opt/anaconda3/envs/ABM_env/lib/python3.10/site-packages/agentpy/tools.py?line=86'>87</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m\u001b[39m__getitem__\u001b[39m(name)\n\u001b[1;32m     <a href='file:///opt/anaconda3/envs/ABM_env/lib/python3.10/site-packages/agentpy/tools.py?line=87'>88</a>\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mKeyError\u001b[39;00m:\n\u001b[1;32m     <a href='file:///opt/anaconda3/envs/ABM_env/lib/python3.10/site-packages/agentpy/tools.py?line=88'>89</a>\u001b[0m     \u001b[39m# Important for pickle to work\u001b[39;00m\n\u001b[0;32m---> <a href='file:///opt/anaconda3/envs/ABM_env/lib/python3.10/site-packages/agentpy/tools.py?line=89'>90</a>\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mAttributeError\u001b[39;00m(name)\n",
      "\u001b[0;31mAttributeError\u001b[0m: variables"
     ]
    }
   ],
   "source": [
    "# Get all logged data for agents\n",
    "results = results.variables.AnimalAgent\n",
    "# Convert the resulting Pandas series to a dataFrame\n",
    "results = results.genome.to_frame().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "'AgentDList' object is not callable",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m/Users/df525/code/weiss_helping_harming2022/helping_harming_model.ipynb Cell 10'\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/df525/code/weiss_helping_harming2022/helping_harming_model.ipynb#ch0000057?line=0'>1</a>\u001b[0m model\u001b[39m.\u001b[39;49mpop(\u001b[39m20\u001b[39;49m)\n",
      "\u001b[0;31mTypeError\u001b[0m: 'AgentDList' object is not callable"
     ]
    }
   ],
   "source": [
    "model.pop()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "probabilities are not non-negative",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m/Users/df525/code/weiss_helping_harming2022/helping_harming_model.ipynb Cell 11'\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/df525/code/weiss_helping_harming2022/helping_harming_model.ipynb#ch0000064?line=0'>1</a>\u001b[0m probs \u001b[39m=\u001b[39m [ p\u001b[39m/\u001b[39m\u001b[39msum\u001b[39m(model\u001b[39m.\u001b[39mpop\u001b[39m.\u001b[39mfitness) \u001b[39mfor\u001b[39;00m p \u001b[39min\u001b[39;00m model\u001b[39m.\u001b[39mpop\u001b[39m.\u001b[39mfitness]\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/df525/code/weiss_helping_harming2022/helping_harming_model.ipynb#ch0000064?line=1'>2</a>\u001b[0m np\u001b[39m.\u001b[39;49mrandom\u001b[39m.\u001b[39;49mchoice(model\u001b[39m.\u001b[39;49mpop,\u001b[39m19\u001b[39;49m,replace\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,p\u001b[39m=\u001b[39;49mprobs)\n",
      "File \u001b[0;32mmtrand.pyx:937\u001b[0m, in \u001b[0;36mnumpy.random.mtrand.RandomState.choice\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: probabilities are not non-negative"
     ]
    }
   ],
   "source": [
    "\n",
    "probs = [ p+model.p.baseline_fitness/sum(model.pop.fitness) for p in model.pop.fitness]\n",
    "np.random.choice(model.pop,19,replace=False,p=probs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1173.9999999999995, 1173.9999999999995, 1179.5, 1178.3999999999999, 1173.9999999999995, 1173.9999999999998, 1284, 1173.9999999999998, 1173.9999999999998, 1284, 1181.6999999999998, 1173.9999999999998, 1284, 1284, 1173.9999999999998, 1173.9999999999998, 1284, 1284, 1284, 1173.9999999999998]"
      ]
     },
     "execution_count": 175,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fitnesses = model.pop.fitness \n",
    "fitnesses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[10.2, -0.8, 10.2, 10.2, 10.2, 10.2, 10.2, 10.2, 10.2, 10.2, 10.2, 10.2, 10.2, 10.2, 10.2, 10.2, 10.2, 10.2, 10.2, 10.2]"
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "probs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[106, 115, 104, 105, 102, 121, 117, 110, 173, 131, 102, 105, 110, 134, 119, 113, 135, 115, 188, 122]"
      ]
     },
     "execution_count": 176,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.pop.age"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "69b725c8c3e55ecc60b7233d0d81a3cfcc6579b2f968ec061419041436e6fb09"
  },
  "kernelspec": {
   "display_name": "Python 3.10.0 ('AMB_env')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
